### DISCUSSION TITLE
Import Form

### DISCUSSION THREAD

**Discussion by insumanth** [vueform/builder#37] [Ideas]
Option to import `.vue` file or `schema` json data to the builder and make edits to the form.

A option to import the form data similar to export will do.
This will help us to make enhancements to existing forms without creating it from scratch.

**Comment by insumanth** [comment]
@adamberecz If we can achieve this using console or other ways, it works for me.

I think this is an important feature to implement as without this, enhancements to existing forms will be very hard.

**Reply by adamberecz** [reply]
If you have a look at the object the builder exports you should be able to create something similar. It should have `form`, `schema`, `theme`, `export` and `builder` properties. You can also learn more about these [here](https://builder.vueform.com/docs/rendering).

The main challenge is to add a new `builder` object for all elements in the schema, which contains extra information for the builder, eg. the exact element type in the builder. Eg. Vueform has `checkboxgroup` element `type` but the builder has multiple alternatives for `checkboxgroup` type:
- `checkboxgroup`
- `checkboxBlocks`
- `checkboxTabs`

These affect what element icon is used in the tree, element selectors and in some cases what kind of extra configuration options it should have, so it's important to set them correctly. You can see all builder element types [here](https://builder.vueform.com/docs/customizing-element-list#available-elements).

If you have an element like this in your manually created schema:
```js
numbers: {
  type: 'checkboxgroup',
  items: [1,2,3],
  view: 'tabs'
}
```

It should be extended like this:
```js
numbers: {
  type: 'checkboxgroup',
  items: [1,2,3],
  view: 'tabs',
  builder: {
    type: 'checkboxTabs',
    json: true,
  }
}
```

Here's what should be provided in the elements' `builder` objects:
- `builder.type` should be set for **each** element according to its type within the builder
- `builder.label` should be set for elements without label (eg. group, divider) - this will be used to for tree and element selectors (eg. in conditions)
- `builder.json = true` should be set for select/multiselect/tags/checkboxgroup/radiogroup elements which options should be displayed as a JSON (in Data/Data source)

Until this gets implemented officially you might create a custom solution based on this and [`load()`](https://builder.vueform.com/docs/loading) manually created JSON schema based forms enriched with `builder` object for each element.

**Reply by letmaik** [reply]
How about just exporting a "proprietary" format that includes all the metadata of the builder and is only meant to be loaded again into the builder. The idea is that it's still fairly easy to copy-paste pieces from the builder into your own app, even if it's not a 1:1 roundtrip.

**Reply by adamberecz** [reply]
That's how it is working right now. I think the question is more about doing it the other way, so if you've already built something manually, without the builder, how you can recreate this "proprietary" format to be able to load it into the builder.

**Reply by letmaik** [reply]
Can you clarify what you mean by that's how it is working right now? I don't see any way to export/download the current "proprietary" format to a file and load it back into the builder at a later point.

**Reply by adamberecz** [reply]
Yes, right. What I meant is that if you have a local copy of the builder you can subscribe to its `@save` event and then you have access to this "proprietary" format that you can eg. store in db and load back later. That's not something possible with the current version of the online builder on our website.

**Reply by bilogic** [reply]
Can have an example? I mean the homepage https://builder.vueform.com does say

> Export JSON
> Use our export function to download exacutable code or save forms via events as a JSON object to database.
